### PR TITLE
38203 Load reCAPTCHA on focus only

### PR DESCRIPTION
**Pull request by dai-eastgate**
related to https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/556

changed log :
Load reCAPTCHA when focus on an input field.

### CODE REVIEW

File: js/onoffice-captchacontrol.js [hunk 1]
```diff
@@ -20,4 +20,21 @@ var onOffice = onOffice || {};
...
+		function addRecaptchaScript() {
+			if (!$("#recaptcha-script").length) {
+				$("<script>")
+					.attr("src", "https://www.google.com/recaptcha/api.js")
+					.attr("async", false)
+					.attr("id", "recaptcha-script")
+					.appendTo("head");
+			}
+		}
+		const targetedFormElements = "#onoffice-form input, #onoffice-form textarea, #onoffice-form button, #onoffice-form";
```
Comment (andernath) [inline js/onoffice-captchacontrol.js:37]: Please use also .oo-forms as selector parent. See onoffice-honeypot.js for reference.

File: plugin/Form.php [hunk 2]
```diff
@@ -600,6 +601,26 @@ public function getEstateContextLabel(): string
...
+			if (wp_script_is('onoffice-captchacontrol', 'registered')) {
+				echo '<script>var requiresCaptchaForm = ' . json_encode($isCaptchaRequired) . ';</script>';
+			} else {
+				wp_register_script('onoffice-captchacontrol', plugins_url('/js/onoffice-captchacontrol.js', ONOFFICE_PLUGIN_DIR . '/index.php'), 'google-recaptcha');
+				add_action('wp_footer', function () use ($isCaptchaRequired) {
+					echo '<script>var requiresCaptchaForm = ' . json_encode($isCaptchaRequired) . ';</script>';
+				});
+			}
+		}
+	}
```
Comment (andernath) [inline plugin/Form.php:623]: Maybe you dont need boolean requiresCaptchaForm, if you use CaptchaDataChecker(oder Handler) to load onoffice-captcha.js only if repatcha is enabled and we are on a form page.

See comment in CaptchaDataChecker.php

File: js/onoffice-captchacontrol.js [hunk 3]
```diff
@@ -20,4 +20,21 @@ var onOffice = onOffice || {};
 
 		return iePosition > 0 || !!navigator.userAgent.match(/Trident.*rv\:11\./);
 	};
-})();
\ No newline at end of file
+})();
+
+var requiresCaptchaForm = requiresCaptchaForm || false;
```
Comment (andernath) [inline js/onoffice-captchacontrol.js:25]: Maybe this boolean is not needed, see comments in Form.php and CaptchaDataChecker.php.

**Review by andernath — COMMENTED** [review]
@dai-eastgate @yeneastgate 
I make some suggestions for structure. But I'm not sure if this is working.

**Comment by dai-eastgate** [comment]
> @dai-eastgate @yeneastgate I make some suggestions for structure. But I'm not sure if this is working.
I will check it. Thanks!

**Comment by dai-eastgate** [comment]
@andernath  I have refactored the code. Please watch my video demo and review the code again. Thanks! 

https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/106214469/0591fbe4-88ff-4867-89e9-027eb0e7152e

**Review by andernath — APPROVED** [review]
Thanks, I will start testing

**Comment by andernath** [comment]
@dai-eastgate 
I got an error testing with custom templates, because the selector in captchacontrol.js is not working for me.
This one:
``` 
const formsSelector = "#onoffice-form, .oo-form";
const inputSelector = "input, textarea, button";
$(`${formsSelector} ${inputSelector}`).on("focus select2:open", addRecaptchaScript);
```

My test template structure causes an error because it is **nested** like:
```
<form method="post" id="onoffice-form">
  <div class="input" style="display: flex; flex-direction: column; align-items: flex-start;">
    <label for="Vorname">Vorname</label>
    <input type="text" name="Vorname" value="">
  </div>
  ...
</form>
```

What do you think about using a selector like 
```
$("#onoffice-form :input, .oo-form :input")
```
to all input, textarea, select and button elements inside of specific forms. (https://api.jquery.com/input-selector/)

**Comment by dai-eastgate** [comment]
> @dai-eastgate I got an error testing with custom templates, because the selector in captchacontrol.js is not working for me. 
OK, I will check and fix it.

**Comment by dai-eastgate** [comment]
@andernath I fixed and tested it, please review again for me. Thanks

**Review by andernath — APPROVED** [review]
Nice. I will retest it.

**Comment by andernath** [comment]
@dai-eastgate 
Sorry I didn't noticed it while reviewing:
At the moment the selector is:
```
 "#onoffice-form, .oo-form :input"
```
Works well for forms with .oo-form but not for old ones with #onoffice-form.
Fix:
```
 "#onoffice-form :input, .oo-form :input"
```

**Comment by dai-eastgate** [comment]
@andernath Let me check again.

**Comment by dai-eastgate** [comment]
> @andernath Let me check again.

@andernath I fixed it. Please watch my video demo. Thanks! 

https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/106214469/de49165e-0b6e-4179-95d0-e2e0cd4079f6

**Review by andernath — DISMISSED** [review]
Next round testing :)

**Review by andernath — APPROVED** [review]
I will test again with fixed master

**Pull request merged**